Parser/myreadline.c

```diff
@@ -373,34 +373,21 @@ PyOS_Readline(FILE *sys_stdin, FILE *sys_stdout, const char *prompt)
     size_t len;
 
     PyThreadState *tstate = _PyThreadState_GET();
-    if (_PyOS_ReadlineTState == tstate) {
+    if (_Py_atomic_load_ptr_relaxed(&_PyOS_ReadlineTState) == tstate) {
         PyErr_SetString(PyExc_RuntimeError,
                         "can't re-enter readline");
         return NULL;
     }
 
-
+    // GH-123321: We need to acquire the lock before setting
+    // _PyOS_ReadlineTState, otherwise the variable may be nullified by a
+    // different thread.
+    PyMutex_Lock(&_PyOS_ReadlineLock);
+    _Py_atomic_store_ptr_relaxed(&_PyOS_ReadlineTState, tstate);
     if (PyOS_ReadlineFunctionPointer == NULL) {
         PyOS_ReadlineFunctionPointer = PyOS_StdioReadline;
     }
 
-    if (_PyOS_ReadlineLock == NULL) {
-        _PyOS_ReadlineLock = PyThread_allocate_lock();
-        if (_PyOS_ReadlineLock == NULL) {
-            PyErr_SetString(PyExc_MemoryError, "can't allocate lock");
-            return NULL;
-        }
-    }
-
-    Py_BEGIN_ALLOW_THREADS
```

Comment:
What about the GIL though? We don't release it for the regular build?

Reply:
Oh, that's right. We don't need it for the `PyMutex`, which handles it internally when it blocks, but we probably still need to release it for the readline call.

Reply:
Does PyMutex take the GIL if it's called outside of the GIL?
Or is it supposed to be called only within the GIL?

Not next to the code atm.

Reply:
You can call `PyMutex_Lock` (and `PyMutex_Unlock`) with or without the GIL.

If the GIL is currently held AND `PyMutex_Lock()` needs to wait on the mutex, it will release the GIL while it waits on the mutex and reacquire the GIL before returning.

If the GIL is not currently held, then it's not touched.